From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: dasd: Fix unimplmentes DIAG function.
References: bnc#643173,LTC#67938
Patch-mainline: Yes

Symptom:     Using a DASD device with the DIAG discipline may lead to
	     a kernel panic when i.e. a reserve operation is executed
	     on this disk from another system.
Problem:     The DASD interrupt handler is called for an unsolicited
	     interrupt and an unimplemented discipline function is called.
Solution:    The informations that can be provided by an unsolicited
	     interrupt are unimportant for a DASD device accessed with
	     the DIAG discipline. Simply ignore the interrupt for DIAG devices.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/block/dasd.c |   21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

--- a/drivers/s390/block/dasd.c
+++ b/drivers/s390/block/dasd.c
@@ -1084,17 +1084,20 @@ void dasd_int_handler(struct ccw_device
 		if (cqr)
 			memcpy(&cqr->irb, irb, sizeof(*irb));
 		device = dasd_device_from_cdev_locked(cdev);
-		if (!IS_ERR(device)) {
-			device->discipline->dump_sense_dbf(device, irb,
-							   "unsolicited");
-			if ((device->features & DASD_FEATURE_ERPLOG))
-				device->discipline->dump_sense(device, cqr,
-							       irb);
-			dasd_device_clear_timer(device);
-			device->discipline->handle_unsolicited_interrupt(device,
-									 irb);
+		if (IS_ERR(device))
+			return;
+		/* ignore unsolicited interrupts for DIAG discipline */
+		if (device->discipline == dasd_diag_discipline_pointer) {
 			dasd_put_device(device);
+			return;
 		}
+		device->discipline->dump_sense_dbf(device, irb,
+						   "unsolicited");
+		if ((device->features & DASD_FEATURE_ERPLOG))
+			device->discipline->dump_sense(device, cqr, irb);
+		dasd_device_clear_timer(device);
+		device->discipline->handle_unsolicited_interrupt(device, irb);
+		dasd_put_device(device);
 		return;
 	}
 
